### PR TITLE
Add cached lazy unmarshalling for field sets

### PR DESCRIPTION
**Pull request by jennybuckley**
**What type of PR is this?**
/kind feature
/wg apply
/sig api-machinery
/priority important-soon
/cc @lavalamp @apelisse @kwiesmueller 

**What this PR does / why we need it**:
This prevents the apiserver from doing any work to re encode a field set that hasn't changed at all during an apply or update. The benefit here is for updates that occur very often but only change one or two fields, like a node heartbeat, etc. This also makes the decoding of the fieldset lazy, which won't give any benefit yet because we always read every fieldset, but in the future we plan to optimize that in structured-merge-diff, and this will make that work possible. Here are some benchmarks:

Before:
```
pkg: k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager
BenchmarkApplyNewObject-12     	    1000	   1674802 ns/op	  500674 B/op	    7167 allocs/op
BenchmarkUpdateNewObject-12    	    1000	   1895343 ns/op	  560238 B/op	    8396 allocs/op
BenchmarkRepeatedUpdate-12     	    2000	    823516 ns/op	  221117 B/op	    2887 allocs/op
```
After
```
pkg: k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager
BenchmarkApplyNewObject-12     	    1000	   1772745 ns/op	  500664 B/op	    7167 allocs/op
BenchmarkUpdateNewObject-12    	    1000	   1823955 ns/op	  558106 B/op	    8362 allocs/op
BenchmarkRepeatedUpdate-12     	    5000	    505358 ns/op	  156538 B/op	    1870 allocs/op
```

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
Part of https://github.com/kubernetes/enhancements/pull/1110

### CODE REVIEW

**Comment by lavalamp** [comment]
This caches the structure but not the serialization, right?

I think we need to hook into the serialization process :/

https://github.com/kubernetes-sigs/structured-merge-diff/pull/90

**Comment by jennybuckley** [comment]
@lavalamp I think we can do both. This is still necessary if we want to avoid deserializing every entry (for example when an update only modifies fields it already manages, we can tell there will be no conflicts early)

/retest

**Comment by lavalamp** [comment]
Not sure I understand-- if we do something to completely defer
serialization for field sets, don't we get this for free?

**Comment by jennybuckley** [comment]
@lavalamp Unless we plan for the interface between the apiserver and smd to be passing in the bytes, I don't think it will defer serialization enough. I had thought that smd#90 was just to expose a faster and more memory efficient serialization/deserialization function for fieldsets, not to defer anything to later

**Comment by lavalamp** [comment]
Yeah, the part that defers it isn't written. I don't think there's a way to
hook this up to the existing structure at all. The goal is to basically
delete all of the code that fills out the existing structure.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go [hunk 1]
```diff
@@ -117,11 +117,39 @@ func decodeVersionedSet(encodedVersionedSet *metav1.ManagedFieldsEntry) (version
...
+	apiVersion fieldpath.APIVersion
+	applied    bool
+}
+
+var _ fieldpath.VersionedSet = &cachedVersionedSet{}
+
+func (c *cachedVersionedSet) Set() *fieldpath.Set {
+	if c.set == nil {
+		set, err := FieldsToSet(c.encodedSet)
+		if err == nil {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go]: That is really unfortunate. This means that if someone changes a managedField in an update, it's not going to fail and just return an empty set, is that what we want?

**Review by apelisse — COMMENTED** [review]
I was also hoping to see the full serialization into a RawExtension/string/[]byte/RawMessage here or in another PR, since we really haven't fixed the read path yet ...

**Comment by apelisse** [comment]
I think we need to:
1. Have Fields be a series of bytes so that it's not deserialized on reads (with a new field name),
2. Use smd#90 to deserialize this series of bytes,
3. Merge this PR so that deserialize when needed, and more importantly, only re-serialize what changed

Technically, with the existing code, READS are still by far the biggest problem we have ...

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go (hunk 1 above)
Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go]: No probably not. We could collect any errors here and return them at the end?

**Comment by jennybuckley** [comment]
@apelisse  I agree that we need to do all 3 of those steps but why does the order matter?

**Comment by jennybuckley** [comment]
>I was also hoping to see the full serialization into a RawExtension/string/[]byte/RawMessage here or in another PR

I think another PR makes more sense

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go (hunk 1 above)
Comment (jennybuckley) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go]: Actually, why not just add an error return value to Set() in the VersionedSet interface. Should be simpler.

**Comment by apelisse** [comment]
Order is not super important indeed, @lavalamp has started a PR for the rawextension. 

Returning an error sounds good

**Comment by jennybuckley** [comment]
I already have one, https://github.com/jennybuckley/kubernetes/commit/e0dd48fc8e9bda41c16c97c8815eb925f1e34004

**Comment by apelisse** [comment]
I'm waiting for `Set()` to return an error and I'll LGTM.

**Comment by apelisse** [comment]
Hey Jenny, I"d love to get this in, do you mind if I follow-up on this?

Thanks

**Comment by jennybuckley** [comment]
@apelisse sure, we just need to be careful of https://github.com/kubernetes/kubernetes/pull/80730#issuecomment-519622953

**Comment by dims** [comment]
/hold

needs rebase

**Comment by apelisse** [comment]
We'll eventually need this, it's not just the highest priority.

**Comment by apelisse** [comment]
@jennybuckley Do we still believe that "we'll eventually need this"?

**Comment by dims** [comment]
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

**Pull request closed without merging**